pkg/apprepo/apprepos_handler.go

```diff
@@ -52,33 +52,56 @@ type combinedClientset struct {
 	*kubernetes.Clientset
 }
 
-// AppRepositoriesHandler handles http requests for operating on app repositories
+// appRepositoriesHandler handles http requests for operating on app repositories
 // in Kubeapps, without exposing implementation details to 3rd party integrations.
-type AppRepositoriesHandler struct {
+type appRepositoriesHandler struct {
 	// The config set internally here cannot be used on its own as a valid
-	// token is required. Call-sites use ConfigForToken to obtain a valid
+	// token is required. Call-sites use configForToken to obtain a valid
 	// config with a specific token.
 	config rest.Config
 
 	// The namespace in which (currently) app repositories are created.
 	kubeappsNamespace string
 
-	// The Kubernetes client using the pod serviceaccount
-	svcKubeClient kubernetes.Interface
+	// clientset using the pod serviceaccount
+	svcClientset combinedClientsetInterface
 
 	// clientsetForConfig is a field on the struct only so it can be switched
-	// for a fake version when testing. NewAppRepositoryHandler sets it to the
+	// for a fake version when testing. NewAppRepositoryhandler sets it to the
 	// proper function below so that production code always has the real
 	// version (and since this is a private struct, external code cannot change
 	// the function).
 	clientsetForConfig func(*rest.Config) (combinedClientsetInterface, error)
+
+	clientset combinedClientsetInterface
+}
+
+type handler interface {
+	CreateAppRepository(appRepoBody io.ReadCloser, requestNamespace string) (*v1alpha1.AppRepository, error)
+	DeleteAppRepository(name, namespace string) error
+	GetNamespaces() ([]corev1.Namespace, error)
+	GetSecret(name, namespace string) (*corev1.Secret, error)
+	GetAppRepository(repoName, repoNamespace string) (*v1alpha1.AppRepository, error)
+}
+
+// AuthHandler exposes handler functionality as a user or the current serviceaccount
+type AuthHandler interface {
+	AsUser(token string) handler
+	AsSVC() handler
 }
 
-// Handler exposes the handler method for testing purposes
-type Handler interface {
-	CreateAppRepository(appRepoBody io.ReadCloser, requestNamespace, token string) (*v1alpha1.AppRepository, error)
-	DeleteAppRepository(name, namespace, token string) error
-	GetNamespaces(token string) ([]corev1.Namespace, error)
+func (a *appRepositoriesHandler) AsUser(token string) handler {
+	clientset, err := a.clientsetForConfig(a.configForToken(token))
+	if err != nil {
+		log.Errorf("unable to create clientset: %v", err)
+	}
+	a.clientset = clientset
+	return a
+}
+
+func (a *appRepositoriesHandler) AsSVC() handler {
+	a.clientset = a.svcClientset
+	return a
```

Comment:
These don't look thread safe? Would it be better to return a copy configured with the correct clientset, rather than updating the internal `a.clientset` for the call-site?

Or, stepping back, it may even make more sense if the current `appRepositoryHandler` was something like a handler factory, that only created a handler when you call `AsUser` or `AsSVC`. Not sure about needing a factory, but the main thing is that it seems in its current form here to be mixing responsibilities... more below.

Reply:
that's a good point, let me check the possibility of modifying the interfaces to fix this and you next comment.